Enabling Macedonian hyphen for 8-bit engines

tstojan opened the pull request:
Enabling Macedonian hyphen for 8-bit engines

1. Adding in `loadhyph-mk.tex`: 
`\input conv-utf8-t2a.tex`
`\input hyph-mk.tex`

2. Removing in `hyph-mk.tex`:
line 192: `1ѓ`
line 359: `2јќ`
line 643: `1ќ`
line 644: `ќа1`
line 645: `2ќ1н`
line 646: `2ќ1т`
containing chars  `ќ (U+045C - cyrkje)` and `ѓ (U+0453 - cyrgje)` that are not in T2A so `conv-utf8-t2a.tex` does not catch them.

In such a way, hyphen for Macedonian would work for both unicode-aware engines (such as XeTeX or LuaTeX) and 8-bit engines (such as TeX or pdfTeX).

mojca commented on hyph-utf8/tex/generic/hyph-utf8/patterns/tex/hyph-mk.tex:
```diff
@@ -189,7 +192,6 @@
 д1с2
 д1т2
 1ду
-1ѓ
```
I guess this is the pattern you had to remove in order to make the patterns loadable in 8-bit encoding?
If so, we could add those patterns into an additional file, we already do that for a bunch of other languages where we duplicate apostrophe etc.

tstojan replied:
Exactly. These 6 patterns (containing `ѓ` and `ќ`) were causing problems once I load:
```
\input conv-utf8-t2a.tex
\input hyph-mk.tex
```
before I removed them from `hyph-mk.tex`. It is not working because chars `ѓ` and `ќ` are not treated in any way in t2a.

I agree with you these 6 patterns can go in a separate file as soon as loading of  `conv-utf8-t2a.tex` and `hyph-mk.tex` do not fail.

mojca commented on hyph-utf8/tex/generic/hyph-utf8/loadhyph/loadhyph-mk.tex:
```diff
@@ -27,7 +27,9 @@
 \else
     % 8-bit engine (such as TeX or pdfTeX)
-    \message{No Macedonian hyphenation patterns - only for Unicode engines}
+    \message{UTF-8 Macedonian hyphenation patterns}
```
This looks semi-OK, but this is a generated file and a complete solution requires changing the sources (written in ruby), so that this file gets generated properly.

After changing the sources a bunch of other files would be created or modified as well (there's a plain text copy of the header, plain text copy of the patterns, there's a file with funny encoding of 8-bit patterns for pTeX, ...)

tstojan replied:
I was not aware they were code-generated. (I noticed there are .tex, .txt with hyphen patters plus other files but since pTex was not enabled (the last branch of if in `loadhyph-mk.tex`), they were not used in Macedonian anyway, so I didn't dare to change something there.) 

I am also OK with your proposal, sounds reasonable, but you probably know better where to edit in this case.

tstojan replied:
Shall we go with accepting this (temporary) solution since changing the ruby files is taking a while?

tstojan commented:
> This change provides everything we need to know, but some minor changes in ruby scripts are needed to automate the rest of the process, only `hyph-mk.tex` changes would actually be needed here, the rest gets auto-generated.

1. Reverted back to the master version of `loadhyph-mk.tex` according to the suggestions. This is auto-generated file. It will be updated automatically once the appropriate ruby files are changed.
2. `hyph-mk.tex` will be the version I updated. @mojca do we still want to add these 6 patterns I removed from `hyph-mk.tex` into a separate file? If yes, could you point me to the right location where such file should sit and to similar files so I can follow guidance and add this file as part of this PR. Or do you maybe prefer to do this when updating ruby files?
3. In such a form, I believe this PR is ready to be approved and this can be followed by changes in the ruby files from your side. Do you have an estimate when this can happen? (I understand that you may follow some development cycles and timings.)

reutenauer commented:
Please have a look at the current structure of the repository. I changed a few things since merging the pull request:

1. Restored the six patterns that use ѓ and ќ in the main file (`hyph-mk.tex`)
1. Added a pattern file for 8-bit engines without the six patterns (`hyph-mk.macedonian.tex`)
1. Adjusted a few settings in the YAML header of the main file to help with automatic generation of the auxiliary files

The Macedonian patterns should now be usable with both UTF-8 and 8-bit engines; the former will see all patterns, the latter only those that can be encoded using the Macedonian-specific 8-bit encoding (i.e. all but six).  The 8-bit pattern file has been generated automatically with the help of the script that generates 8-bit files for pTeX, hence the reference to pTeX in that file.  It also uses the caret escape mechanism to encode all non-ASCII characters.  This is somewhat unwieldy, but can be adjusted manually since that file is going to be static in the future.

The slightly weird message “MACEDONIAN Macedonian hyphenation patterns” displayed by the loader is also automatically generated, and is explained by the fact that the first “MACEDONIAN” is the encoding name, hence the capitalisation.  This can obviously also be adjusted; the simplest solution may actually be to change the encoding name :-)

tstojan commented:
Thanks @reutenauer . Having “MACEDONIAN as encoding name is just fine.

I now see issues when running pdflatex or pdftex, so it doesn't work. It does not show the message:  “MACEDONIAN Macedonian hyphenation patterns” (it does not enter the appropriate if branch in `loadhyph-mk.tex`) and it is stuck in a pdf compile that never finishes for MikTeX and for Texlive it does not do updates.

Here is a minimal example you can try compiling it and see whether it produced hyphened text:

```
\documentclass{article}
\usepackage[macedonian,showlanguages]{babel}

\begin{document}

\section{Примери со делење на слогови (hyphen)}

\subsection{Генетика (извор: македонска википедија)}
Генетика (од старогрчки: генетикос, да роди и од генезис, потекло) или наследознание — гранка на биологијата која се занимава со проучување на гените, наследноста и варијацијата кај живите организми.

Генетиката се справува со прашања какви што се молекуларната градба и функција на гените, однесувањето на гените во контекст на клетка или организам (на пример, доминантност и епигенетика), закономерностите на наследување од родителите до потомството, како и генската дистрибуција, варијација и промена во популациите. Поради тоа што гените се универзални за сите живи организми, генетиката може да се користи за проучување на сите живи системи, од вируси до бактерии, преку растенијата и домашните животни, до човекот (како што е во медицинската генетика). Природно ДНК се среќава кај коњот.

\end{document}
```

reutenauer commented:
Did you try uploading TeX Live?  This won’t work.  I haven’t uploaded the code to CTAN yet as I was hoping you could test it first from this repository.

tstojan commented:
@reutenauer : I installed the ruby bundle, this went without errors, I see the dedicated files are in `loadhyph-mk.tex`, `hyph-mk.macedonian.tex` are all in texlive in the right folders, but this is what I get as a printout now:

```
> pdflatex example_hyphen_mk.tex
This is pdfTeX, Version 3.14159265-2.6-1.40.21 (TeX Live 2020/W32TeX) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./example_hyphen_mk.tex
LaTeX2e <2020-02-02> patch level 5
L3 programming layer <2020-05-15>
(c:/texlive/2020/texmf-dist/tex/latex/base/article.cls
Document Class: article 2019/12/20 v1.4l Standard LaTeX document class
(c:/texlive/2020/texmf-dist/tex/latex/base/size10.clo))
(c:/texlive/2020/texmf-dist/tex/generic/babel/babel.sty
(c:/texlive/2020/texmf-dist/tex/generic/babel/babel.def
(c:/texlive/2020/texmf-dist/tex/generic/babel/txtbabel.def))
(c:/texlive/2020/texmf-dist/tex/generic/babel-macedonian/macedonian.ldf
(c:/texlive/2020/texmf-dist/tex/latex/cyrillic/t2aenc.def
(c:/texlive/2020/texmf-dist/tex/latex/base/t2aenc.dfu))

Package babel Warning: No input encoding specified for Macedonian language on i
nput line 155.
```

A pdf is produced but there is no hyphening and there is row overflow.
 
I don't see the message “MACEDONIAN Macedonian hyphenation patterns” (so it is likely not entering the if branch). Could you try compiling the minimum working example from above on your side?

tstojan commented:
`bundle` works for me, but `bundle exec rake` fails with some ruby errors like:

```
hyph-utf8/source/generic/hyph-utf8/generate-plain-patterns.rb
Traceback (most recent call last):
        3: from hyph-utf8/source/generic/hyph-utf8/generate-plain-patterns.rb:7:in `<main>'
        2: from hyph-utf8/source/generic/hyph-utf8/generate-plain-patterns.rb:7:in `require'
        1: from C:/Ruby26/lib/ruby/gems/2.6.0/gems/unicode-0.4.4.4-x86-mingw32/lib/unicode.rb:2:in `<top (required)>'
C:/Ruby26/lib/ruby/gems/2.6.0/gems/unicode-0.4.4.4-x86-mingw32/lib/unicode.rb:2:in `require': cannot load such file -- unicode/2.6/unicode_native (LoadError)
        4: from hyph-utf8/source/generic/hyph-utf8/generate-plain-patterns.rb:7:in `<main>'
        3: from hyph-utf8/source/generic/hyph-utf8/generate-plain-patterns.rb:7:in `require'
        2: from C:/Ruby26/lib/ruby/gems/2.6.0/gems/unicode-0.4.4.4-x86-mingw32/lib/unicode.rb:1:in `<top (required)>'
        1: from C:/Ruby26/lib/ruby/gems/2.6.0/gems/unicode-0.4.4.4-x86-mingw32/lib/unicode.rb:5:in `rescue in <top (required)>'
```
Anyhow, I am OK with moving it into CTAN: if it works on your for 8-bit cases with some minimum working example it is probably OK, in any case once in CTAN I can test it additionally and tell you, if there are some issues.
